LFS-1067: Themed 404 not found page

veronikaslc opened the pull request:
Navigate to `http://localhost:8080/co` to see the custom 404 page.

Note: the mockup in the jira issue is no longer valid, we switched "404 not found" from accent color to the theme's primary color.

Commit:
LFS-1067: Themed 404 not found page

sdumitriu commented on modules/commons/pom.xml:
```diff
@@ -42,6 +42,7 @@
             <!-- Initial content to be loaded on bundle installation -->
             <Sling-Initial-Content>
               SLING-INF/content/libs/lfs/Resource/;path:=/libs/lfs/Resource/;overwrite:=true,
+              SLING-INF/content/libs/slings/errorhandler/;path:=/libs/slings/errorhandler/;overwrite:=true,
```
```suggestion
              SLING-INF/content/libs/sling/servlet/errorhandler/;path:=/libs/sling/servlet/errorhandler/;overwrite:=true,
```

Also move the jsp file.

veronikaslc committed:
LFS-1067: Themed 404 not found page

Correct the path in pom.xml

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

Commit:
LFS-1067: Themed 404 not found page

sashaandjic reviewed (approved):
tested

after building this branch, 404.jsp was already in place 
![image](https://user-images.githubusercontent.com/23243244/114946908-14309a80-9e1a-11eb-8bb3-509363bcadde.png)


![image](https://user-images.githubusercontent.com/23243244/114946844-f5ca9f00-9e19-11eb-807b-f9fe2f0c3e9e.png)

![image](https://user-images.githubusercontent.com/23243244/114946825-ec413700-9e19-11eb-81b6-9def16b39867.png)

![image](https://user-images.githubusercontent.com/23243244/114946809-e5b2bf80-9e19-11eb-81fb-751b49f86fdf.png)

Commit:
LFS-1067: Themed 404 not found page
Refactoring

sdumitriu committed:
LFS-1067: Themed 404 not found page
Removed broken header/footer inclusion (this template already has the header and footer in place

sdumitriu committed:
LFS-1067: Themed 404 not found page
Set the right content type

sdumitriu committed:
LFS-1067: Themed 404 not found page
Fixed JS error

Commit:
LFS-1067: Themed 404 not found page
Fixed redirect button and styles

Commit:
LFS-1067: Themed 404 not found page
Separated styles for 404 page

marta- committed:
LFS-1067: Themed 404 not found page
Replaced "Go to dashboard" button with an extended Fab, per mockup

marta- commented on modules/commons/src/main/frontend/src/notFoundStyles.js:
```diff
@@ -0,0 +1,46 @@
+//
+import { lightBlue } from '@material-ui/core/colors';
+
+const styles = theme => ({
```
Since these styles are not shared with other components, they don't need to be separated from the component's code.

marta- commented on modules/commons/src/main/frontend/src/components/Component404.js:
```diff
@@ -0,0 +1,82 @@
+
+import styles from "../notFoundStyles";
+
+class Container404 extends React.Component {
```
Please refactor as a functional component.

Some of our initial code is class based awaiting refactoring, but all new components should be functional.

I would also rename this as `PageNotFound`.

Commit:
LFS-1067: Themed 404 not found page
Moved style into component
Made component functional and renamed

marta- commented on modules/commons/src/main/frontend/src/components/PageNotFound.js:
```diff
@@ -0,0 +1,103 @@
+    maxWidth: "200px",
+  },
+  notFoundTitle: {
+    color: lightBlue.A400,
```
This was meant to be the theme's accent color: blue for LFS, orange for cardiac_rehab, etc. Since it's currently not that easy to access the accent color, let's instead use `theme.palette.primary.main` here instead.

marta- commented on modules/commons/src/main/frontend/src/components/PageNotFound.js:
```diff
@@ -0,0 +1,103 @@
+  },
+  notFoundMessage: {
+    paddingTop: theme.spacing(1),
+    paddingBottom: theme.spacing(2)
```
`color: theme.palette.textSecondary` would work better than the default text color now that we're changing the color of the title.

Commit:
LFS-1067: Themed 404 not found page
Better text coloring

marta- committed:
LFS-1067: Themed 404 not found page
Simplified styling

sashaandjic commented:
I tried to access form that was deleted in another tab and got this:
![image](https://user-images.githubusercontent.com/23243244/115848607-0b2f6280-a3f2-11eb-9ee2-894481e2d82e.png)

However, I was expecting this:
![image](https://user-images.githubusercontent.com/23243244/115848680-1e423280-a3f2-11eb-944a-d235d4b43b28.png)

sdumitriu committed:
LFS-1067: Themed 404 not found page
Set the right status code

sdumitriu commented:
@sashaandjic it's a bug and a non-implemented feature.

The bug is that there should be a message after the colon, like there used to be before. And I did fix that bug, now we're back to displaying "Error obtaining form data: 404 Not Found".

The non-implemented feature is that at the moment the custom 404 page is only displayed when accessing a non-existing URL without the `/content.html` prefix, but not inline. As future improvements not to be addressed in this PR:

- also display the pretty message inline when accessing through the content UI
- maybe have a custom JSON response when accessing a `.json` URL instead of HTML?

marta- commented:
> 
> The non-implemented feature is that at the moment the custom 404 page is only displayed when accessing a non-existing URL without the `/content.html` prefix, but not inline. 

Agreed, I had this lined up as a new task (not yet reported).
